<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [1]:
### YOUR CODE STARTS HERE
#load data
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-09-17 04:43:57--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2019-09-17 04:44:03 (639 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [2]:
# Load Data
import pandas as pd
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [3]:
# clean data

import numpy as np
df = df.replace({'?':np.NaN, 'n':0, 'y':1})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
# Filter the dataframe into a democrats and republicans dataframes

dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']

In [5]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [6]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [7]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

# 1 sample T-test

In [0]:
from scipy.stats import ttest_1samp

1) Null Hypothesis:

In 1-sample t-tests YOU GET TO CHOOSE YOUR NULL HYPOTHESIS

𝐻0 : 0.0 - There is ZERO republican support for this bill

2) Alternative Hypothesis

𝐻𝑎 :  𝑥¯≠0  - There is non-zero support for the budget bill among repulbicans.

3) Confidence Level: 95% or .95

In [9]:
# Run 1-sample t-test providing sample and null hypothesis
# pass nan_policy='omit' any time you nave NaN values in a column
ttest_1samp(rep['budget'], 0, nan_policy='omit')


Ttest_1sampResult(statistic=5.02528816355446, pvalue=1.3107521045196396e-06)

4) t-statistic: 5.025

5) p-value of .00000131

------------

p values is very very low. so I reject the null hypothesis.

In [10]:
rep['budget'].mean()

0.13414634146341464

## 2 Sample T-test


In [0]:
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

In [12]:
# This is a powerful library, there's a lot that we can do with it
import scipy.stats
dir(scipy.stats)

['PearsonRConstantInputWarning',
 'PearsonRNearConstantInputWarning',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_binned_statistic',
 '_constants',
 '_continuous_distns',
 '_discrete_distns',
 '_distn_infrastructure',
 '_distr_params',
 '_hypotests',
 '_multivariate',
 '_rvs_sampling',
 '_stats',
 '_stats_mstats_common',
 '_tukeylambda_stats',
 'absolute_import',
 'alpha',
 'anderson',
 'anderson_ksamp',
 'anglit',
 'ansari',
 'arcsine',
 'argus',
 'bartlett',
 'bayes_mvs',
 'bernoulli',
 'beta',
 'betaprime',
 'binned_statistic',
 'binned_statistic_2d',
 'binned_statistic_dd',
 'binom',
 'binom_test',
 'boltzmann',
 'boxcox',
 'boxcox_llf',
 'boxcox_normmax',
 'boxcox_normplot',
 'bradford',
 'brunnermunzel',
 'burr',
 'burr12',
 'cauchy',
 'chi',
 'chi2',
 'chi2_contingency',
 'chisquare',
 'circmean',
 'circstd',
 'circvar',
 'combine_pvalues',
 'contingency',
 'cosine',
 'crystalball',

###  p > 0.1 
#### water-project
their mean shows that they are almost the same.

### p < 0.1

#### handicapped-infants
#### budget
dem support more than the rep


### p < 0.1

rep support more than the rep


In [13]:
# budget

ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

In [14]:
rep['budget'].mean()

0.13414634146341464

In [15]:
dem['budget'].mean()

0.8884615384615384

In [17]:
# water-project
ttest_ind(rep['water-project'],dem['water-project'],nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [18]:
rep['water-project'].mean()

0.5067567567567568

In [19]:
dem['water-project'].mean()

0.502092050209205

In [23]:
# 'handicapped-infants'
ttest_ind(rep['handicapped-infants'],dem['handicapped-infants'],nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [21]:
rep['handicapped-infants'].mean()

0.18787878787878787

In [22]:
dem['handicapped-infants'].mean()

0.6046511627906976

In [0]:
ttest_ind(rep['handicapped-infants'],dem['handicapped-infants'],nan_policy='omit')

In [24]:
# physician-fee-freeze
ttest_ind(rep['physician-fee-freeze'],dem['physician-fee-freeze'],nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

In [25]:
rep['physician-fee-freeze'].mean()

0.9878787878787879

In [26]:
dem['physician-fee-freeze'].mean()

0.05405405405405406

In [27]:
#el-salvador-aid
ttest_ind(rep['el-salvador-aid'],dem['el-salvador-aid'],nan_policy='omit')

Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)

In [28]:
rep['el-salvador-aid'].mean()

0.9515151515151515

In [29]:
dem['el-salvador-aid'].mean()

0.21568627450980393

In [30]:
#religious-groups
ttest_ind(rep['religious-groups'],dem['religious-groups'],nan_policy='omit')

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

In [31]:
rep['religious-groups'].mean()

0.8975903614457831

In [32]:
dem['religious-groups'].mean()

0.47674418604651164

In [33]:
# anti-satellite-ban
ttest_ind(rep['anti-satellite-ban'],dem['anti-satellite-ban'],nan_policy='omit')

Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)

In [34]:
rep['anti-satellite-ban'].mean()

0.24074074074074073

In [35]:
dem['anti-satellite-ban'].mean()

0.7722007722007722

#### handicapped-infants   -  p < 0.1 - dem > rep 
#### water-project         -  p > 0.1 - dem ~ rep
#### budget                -  p < 0.1 - dem > rep 
#### physician-fee-freeze  -  p < 0.1 - dem < rep
#### el-salvador-aid       -  p < 0.1 - dem < rep
#### religious-groups      -  p < 0.1 - dem < rep
#### anti-satellite-ban    -  p < 0.1 - dem > rep
#### aid-to-contras
#### mx-missile
#### immigration
#### synfuels
#### education
#### right-to-sue
#### crime
#### duty-free
#### south-africa